src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -113,10 +121,32 @@ public bool TryGetProjectState(ProjectId projectId, [NotNullWhen(true)] out Proj
                 => _projectStates.TryGetValue(projectId, out state);
 
             public ActiveFileState GetOrCreateActiveFileState(DocumentId documentId)
-                => _activeFileStates.GetOrAdd(documentId, id => new ActiveFileState(id));
+            {
+                if (!_activeFileStates.TryGetValue(documentId, out var value))
+                {
+                    lock (_lock)
+                    {
+                        value = new ActiveFileState(documentId);
+                        _activeFileStates = _activeFileStates.SetItem(documentId, value);
```

Comment:
this feels racey.  if two things call into GetOrCreateActiveFileState.

I believe we have IMmutableInterlocked to help with this pattern.

Reply:
Did the double check as we discussed here and below.